File: languages/python/k-means.py
import sys

P = 1000000007
K = 16
D = 4
ITERS = 10
RANGE = 256


def k_means(n):
    # 1. Generate N integer D-dimensional points with the pinned LCG
    pt = [0] * (n * D)
    state = 42
    for i in range(n * D):
        state = (state * 1103515245 + 12345) & 0x7FFFFFFF
        pt[i] = state % RANGE
    cen = pt[: K * D]                       # initial centroids = first K points
    assign = [0] * n

    # 2. ITERS iterations of assign + update
    for _ in range(ITERS):
        for i in range(n):                  # assignment - nearest centroid
            base = i * D
            best = 0
            bd = -1
            for k in range(K):
                kb = k * D
                dist = 0
                for d in range(D):
                    df = pt[base + d] - cen[kb + d]
                    dist += df * df
                if bd < 0 or dist < bd:     # STRICT < : ties go to the lowest k
                    bd = dist
                    best = k
            assign[i] = best
        ssum = [0] * (K * D)                # update - floor-mean, empty unchanged
        cnt = [0] * K
        for i in range(n):
            k = assign[i]
            cnt[k] += 1
            base = i * D
            kb = k * D
            for d in range(D):
                ssum[kb + d] += pt[base + d]
        for k in range(K):
            if cnt[k] > 0:
                kb = k * D
                c = cnt[k]
                for d in range(D):
                    cen[kb + d] = ssum[kb + d] // c   # INTEGER (floor) division

    for i in range(n):                      # final assignment with final centroids
        base = i * D
        best = 0
        bd = -1
        for k in range(K):
            kb = k * D
            dist = 0
            for d in range(D):
                df = pt[base + d] - cen[kb + d]
                dist += df * df
            if bd < 0 or dist < bd:
                bd = dist
                best = k
        assign[i] = best

    h = 0
    for v in cen:
        h = (h * 31 + v) % P
    for i in range(n):
        h = (h * 31 + assign[i]) % P
    return h


if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 8000
    print(k_means(n))
    print("k-means(%d)" % n)

File: languages/javascript/k-means.js
"use strict";

// k-means: hand-written integer Lloyd's clustering (K=16, D=4, ITERS=10). See
// sort-search.js for why the LCG uses Math.imul.

const P = 1000000007;
const K = 16;
const D = 4;
const ITERS = 10;
const RANGE = 256;

function lcgNext(s) {
  return (Math.imul(s, 1103515245) + 12345) & 0x7fffffff;
}

function kMeans(n) {
  // 1. Generate N integer D-dimensional points with the pinned LCG
  const pt = new Int32Array(n * D);
  let state = 42;
  for (let i = 0; i < n * D; i++) {
    state = lcgNext(state);
    pt[i] = state % RANGE;
  }
  const cen = new Int32Array(K * D);
  for (let i = 0; i < K * D; i++) cen[i] = pt[i]; // initial centroids = first K points
  const assign = new Int32Array(n);

  // 2. ITERS iterations of assign + update
  for (let iter = 0; iter < ITERS; iter++) {
    for (let i = 0; i < n; i++) {
      // assignment - nearest centroid
      const base = i * D;
      let best = 0;
      let bd = -1;
      for (let k = 0; k < K; k++) {
        const kb = k * D;
        let dist = 0;
        for (let d = 0; d < D; d++) {
          const df = pt[base + d] - cen[kb + d];
          dist += df * df;
        }
        if (bd < 0 || dist < bd) {
          // STRICT < : ties go to the lowest k
          bd = dist;
          best = k;
        }
      }
      assign[i] = best;
    }
    const ssum = new Int32Array(K * D); // update - floor-mean, empty unchanged
    const cnt = new Int32Array(K);
    for (let i = 0; i < n; i++) {
      const k = assign[i];
      cnt[k] += 1;
      const base = i * D;
      const kb = k * D;
      for (let d = 0; d < D; d++) {
        ssum[kb + d] += pt[base + d];
      }
    }
    for (let k = 0; k < K; k++) {
      if (cnt[k] > 0) {
        const kb = k * D;
        const c = cnt[k];
        for (let d = 0; d < D; d++) {
          cen[kb + d] = Math.floor(ssum[kb + d] / c); // INTEGER (floor) division
        }
      }
    }
  }

  for (let i = 0; i < n; i++) {
    // final assignment with final centroids
    const base = i * D;
    let best = 0;
    let bd = -1;
    for (let k = 0; k < K; k++) {
      const kb = k * D;
      let dist = 0;
      for (let d = 0; d < D; d++) {
        const df = pt[base + d] - cen[kb + d];
        dist += df * df;
      }
      if (bd < 0 || dist < bd) {
        bd = dist;
        best = k;
      }
    }
    assign[i] = best;
  }

  let h = 0;
  for (let v = 0; v < K * D; v++) {
    h = (h * 31 + cen[v]) % P;
  }
  for (let i = 0; i < n; i++) {
    h = (h * 31 + assign[i]) % P;
  }
  return h;
}

function main() {
  const n = process.argv[2] !== undefined ? parseInt(process.argv[2], 10) : 8000;
  console.log(kMeans(n));
  console.log(`k-means(${n})`);
}

main();
